### PR TITLE
Introduce metered container

### PR DESCRIPTION
**Pull request by MathieuDutSik**
## Motivation

When looking for the performance of the storage, the solution is a metered wrapper. This is introduced
here and implemented for RocksDb, ScyllaDb and DynamoDb.

## Proposal

The following was done:
* Introduction of the MeteredStore and proof of conservation of ReadableKeyValueStore, etc.
* Unfortunately, I could not address the DirectWritableKeyValueStore because of a mysterious lifetime issue.
* Implementation for RocksDb, ScyllaDb, DynamoDb. It is controlled by the "metrics" feature
* The metric is done for the original container, but also for the LruCaching and ValueSplitting ones.

## Test Plan

The CI does not change, but the use has to be checked.

## Release Plan

That should be used by developer mostly.

## Links

- [reviewer checklist](https://github.com/linera-io/linera-protocol/blob/main/CONTRIBUTING.md#reviewer-checklist)

### CODE REVIEW

File: linera-views/src/metered_wrapper.rs [hunk 1]
```diff
@@ -0,0 +1,217 @@
...
+    read_value_bytes: HistogramVec,
+    contains_key: HistogramVec,
+    read_multi_values_bytes: HistogramVec,
+    find_keys_by_prefix: HistogramVec,
+    find_key_values_by_prefix: HistogramVec,
+    write_batch: HistogramVec,
+    clear_journal: HistogramVec,
+}
+
+/// The metered counter for the "rocks db"
```
Comment (ma2bd) [inline linera-views/src/metered_wrapper.rs:29]: See previous comment. We should probably replace `(metered) counter` by `metrics` (plural) everywhere.

**Review by ma2bd — COMMENTED** [review]
Nice! Just a few comments on names

File: linera-views/src/metered_wrapper.rs [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+/// A metered wrapper that keeps track of every operation
+#[derive(Clone)]
+pub struct MeteredStore<K> {
+    /// the metrics being stored
+    counter: MeteredCounter,
+    /// The underlying store of the metered store
+    pub store: K,
+}
+
+async fn prometheus_async<F, O>(f: F, hist: &HistogramVec) -> O
```
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-views/src/metered_wrapper.rs:119]: Ok.

File: linera-views/src/metered_wrapper.rs [hunk 3]
```diff
@@ -9,12 +9,14 @@ use crate::{
...
 use async_trait::async_trait;
 use convert_case::{Case, Casing};
+use linera_base::sync::Lazy;
 use prometheus::{register_histogram_vec, HistogramVec};
 use std::{future::Future, time::Instant};
 //use crate::journaling::DirectWritableKeyValueStore;
 
 #[derive(Clone)]
-struct MeteredCounter {
+/// The implementation of the `MeteredCounter` for the `KeyValueStore`.
```
[the remark this answers is not in the document]
Comment (MathieuDutSik) [reply linera-views/src/metered_wrapper.rs:18]: Rename as `KeyValueStoreMetrics`.

File: linera-views/src/metered_wrapper.rs (hunk 1 above)
Comment (MathieuDutSik) [reply linera-views/src/metered_wrapper.rs]: Ok, rename as `metrics`.

**Comment by ma2bd** [comment]
@MathieuDutSik Still in the summary:
> Unfortunately, I could not address the DirectWritableKeyValueStore because of a mysterious lifetime issue.

I believed you solved this issue, right?

File: linera-views/src/dynamo_db.rs [hunk 4]
```diff
@@ -40,6 +40,12 @@ use linera_base::ensure;
 use std::{collections::HashMap, env, str::FromStr, sync::Arc};
 use thiserror::Error;
 
+#[cfg(feature = "metrics")]
+use crate::metering::{
+    MeteredStore, METERED_COUNTER_DYNAMO_DB, METERED_COUNTER_LRU_CACHING,
```
Comment (ma2bd) [inline linera-views/src/dynamo_db.rs:45]: A few more to be renamed: `METERED_COUNTER_DYNAMO_DB` -> `DYNAMO_DB_METRICS` etc

File: linera-views/src/metering.rs [hunk 5]
```diff
@@ -0,0 +1,218 @@
+// Copyright (c) Zefchain Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+//use std::time::{Duration, Instant};
+//use linera_base::sync::Lazy;
```
Comment (ma2bd) [inline linera-views/src/metering.rs:5]: remove

**Comment by MathieuDutSik** [comment]
> @MathieuDutSik Still in the summary:
> 
> > Unfortunately, I could not address the DirectWritableKeyValueStore because of a mysterious lifetime issue.
> 
> I believed you solved this issue, right?

No, I did not. However, I consider that this is not blocking for merging the PR.

File: linera-views/src/dynamo_db.rs (hunk 4 above)
Comment (MathieuDutSik) [reply linera-views/src/dynamo_db.rs]: Ok, done.

File: linera-views/src/metering.rs (hunk 5 above)
Comment (MathieuDutSik) [reply linera-views/src/metering.rs]: Done;

**Review by ma2bd — APPROVED** [review]
Thanks!

**Pull request merged**